shiny/ui/dataframe/js/selection.tsx

```diff
@@ -0,0 +1,180 @@
+  mode: SelectionMode,
+  keyAccessor: (el: TElement) => TKey,
+  focusOffset: (start: TKey, offset: number) => TKey,
+  between?: (from: TKey, to: TKey) => ReadonlyArray<TKey>
+): SelectionSet<TKey, TElement> {
+  const [selectedKeys, setSelectedKeys] = useState<ImmutableSet<TKey>>(
+    ImmutableSet.empty()
+  );
+
+  // The anchor is the item that was most recently selected with a click or ctrl-click,
+  // and is used to determine the "other end" of a shift-click selection operation.
+  const [anchor, setAnchor] = useState<TKey | null>(null);
+
+  const onMouseDown = (event: React.MouseEvent<TElement, MouseEvent>): void => {
+    if (mode === SelectionMode.None) {
+      return;
+    }
+
+    const el = event.currentTarget as TElement;
+    const key = keyAccessor(el);
+
+    const result = performMouseDownAction<TKey, TElement>(
+      mode,
+      between,
+      selectedKeys,
+      event,
+      key,
+      anchor
+    );
+    if (result) {
+      setSelectedKeys(result.selection);
+      if (result.anchor) {
+        setAnchor(key);
+        el.focus();
+      }
+      event.preventDefault();
+    }
+  };
+
+  const onKeyDown = (event: React.KeyboardEvent<TElement>): void => {
+    if (mode === SelectionMode.None) {
+      return;
+    }
+
+    const el = event.currentTarget as TElement;
+    const key = keyAccessor(el);
+    const selected = selectedKeys.has(key);
+
+    if (mode === SelectionMode.Single) {
+      if (event.key === " " || event.key === "Enter") {
+        if (selectedKeys.has(key)) {
+          setSelectedKeys(ImmutableSet.empty());
+        } else {
+          setSelectedKeys(ImmutableSet.just(key));
+        }
+        event.preventDefault();
+      } else if (event.key === "ArrowUp" || event.key === "ArrowDown") {
+        const targetKey = focusOffset(key, event.key === "ArrowUp" ? -1 : 1);
+        if (targetKey) {
+          event.preventDefault();
+          if (selected) {
+            setSelectedKeys(ImmutableSet.just(targetKey));
+          }
+        }
+      }
+    } else if (mode === SelectionMode.MultiToggle) {
+      if (event.key === " " || event.key === "Enter") {
+        setSelectedKeys(selectedKeys.toggle(key));
+        event.preventDefault();
+      } else if (event.key === "ArrowUp" || event.key === "ArrowDown") {
+        if (focusOffset(key, event.key === "ArrowUp" ? -1 : 1)) {
+          event.preventDefault();
+        }
+      }
+    }
+  };
+
+  return {
+    has(key: TKey): boolean {
+      return selectedKeys.has(key);
+    },
+
+    set(key: TKey, selected: boolean) {
+      if (selected) {
+        setSelectedKeys(selectedKeys.add(key));
+      } else {
+        setSelectedKeys(selectedKeys.delete(key));
+      }
+    },
+
+    clear() {
+      setSelectedKeys(selectedKeys.clear());
+    },
+
+    keys() {
+      return selectedKeys;
+    },
+
+    itemHandlers() {
+      return { onMouseDown, onKeyDown };
+    },
+  };
+}
+
+const isMac = /^Mac/.test(window.navigator.platform);
```

Comment:
Apparently `window.navigator.platform` is deprecated. https://developer.mozilla.org/en-US/docs/Web/API/Navigator/platform

https://stackoverflow.com/questions/10527983/best-way-to-detect-mac-os-x-or-windows-computers-with-javascript-or-jquery

Reply:
WHY 😠